piker/brokers/kucoin.py

```diff
@@ -0,0 +1,668 @@
+                break
+
+        # Map to OHLC values to dict then to np array
+        new_bars = []
+        for i, bar in enumerate(bars[::-1]):
+            data = {
+                'index': i,
+                'time': bar[0],
+                'open': bar[1],
+                'close': bar[2],
+                'high': bar[3],
+                'low': bar[4],
+                'volume': bar[5],
+                'amount': bar[6],
+                'bar_wap': 0.0,
+            }
+
+            row = []
+            for _, (field_name, field_type) in enumerate(_ohlc_dtype):
+                value = data[field_name]
+
+                match field_name:
+                    case 'index':
+                        row.append(int(value))
+                    case 'time':
+                        row.append(value)
+                    case _:
+                        row.append(float(value))
+
+            new_bars.append(tuple(row))
+
+        array = np.array(new_bars, dtype=_ohlc_dtype) if as_np else bars
+        return array
+
+
+def fqsn_to_kucoin_sym(fqsn: str, pairs: dict[str, KucoinMktPair]) -> str:
+    pair_data = pairs[fqsn]
+    return pair_data.baseCurrency + '-' + pair_data.quoteCurrency
+
+
+def kucoin_sym_to_fqsn(sym: str) -> str:
+    return sym.lower().replace('-', '')
+
+
+@acm
+async def get_client() -> AsyncGenerator[Client, None]:
+    client = Client()
+    await client.cache_pairs()
+
+    yield client
+
+
+@tractor.context
+async def open_symbol_search(
+    ctx: tractor.Context,
+) -> None:
+    async with open_cached_client('kucoin') as client:
+        # load all symbols locally for fast search
+        await client.cache_pairs()
+        await ctx.started()
+
+        async with ctx.open_stream() as stream:
+            async for pattern in stream:
+                await stream.send(await client.search_symbols(pattern))
+                log.info('Kucoin symbol search opened')
+
+
+async def stream_quotes(
+    send_chan: trio.abc.SendChannel,
+    symbols: list[str],
+    feed_is_live: trio.Event,
+    loglevel: str = None,
+    # startup sync
+    task_status: TaskStatus[tuple[dict, dict]] = trio.TASK_STATUS_IGNORED,
+) -> None:
+    '''
+    Required piker api to stream real-time data.
+    Where the rubber hits the road baby
+
+    '''
+    connect_id = str(uuid4())
+
+    async with open_cached_client('kucoin') as client:
+        log.info('Starting up quote stream')
+        # loop through symbols and sub to feedz
+        for sym in symbols:
+            token, ping_interval = await client._get_ws_token()
+            pairs = await client.cache_pairs()
+            kucoin_sym = pairs[sym].symbol
+
+            init_msgs = {
+                # pass back token, and bool, signalling if we're the writer
+                # and that history has been written
+                sym: {
+                    'symbol_info': {
+                        'asset_type': 'crypto',
+                        'price_tick_size': 0.0005,
```

Comment:
wait why are you hardcoding this instead of pulling it from the pair info in `pairs`?

Reply:
I'm actually unsure of what the `price_tick_size` and `lot_tick_size` are. Do they map with any values in `KucoinMktPair`?

Reply:
@goodboy - I think we need better documentation on init message formatting.

Reply:
the short answer is yes but also this is being more formalized in #489 .. unfortunately a bit later then i would have liked 😂 

if you wanted to get ambitious you could try to adopt the stuff in there, but it's probably fine to get changed later.

more or less `price/lot_tick_size` are the float/Decimal values which specify the *min increment value* for both the market's price and volume/size.

Reply:
so yes, you should pull them from `KucoinMktPair`.

likely the two values inline here:
https://github.com/pikers/piker/pull/494/files#diff-0e687c4074ef81501ccba2717591ca55647c53b9b3dc85c33d4b4fc1436c644cR78